src/runtime.js

```diff
@@ -59,15 +59,19 @@ var RuntimeGenerator = {
 
   // allocation on the top of memory, adjusted dynamically by sbrk
   dynamicAlloc: function(size) {
-    if (ASSERTIONS) size = '(assert(DYNAMICTOP > 0),' + size + ')'; // dynamic area must be ready
-#if USE_PTHREADS
-    if (typeof ENVIRONMENT_IS_PTHREAD !== 'undefined' && ENVIRONMENT_IS_PTHREAD) throw 'Runtime.dynamicAlloc is not available in pthreads!'; // This is because each worker has its own copy of DYNAMICTOP, of which main thread is authoritative.
-#endif
-    var ret = RuntimeGenerator.alloc(size, 'DYNAMIC');
-    if (SAFE_HEAP) ret += '; if (asm) { Runtime.setDynamicTop(DYNAMICTOP); }';
-    ret += '; if (DYNAMICTOP >= TOTAL_MEMORY) { var success = enlargeMemory(); if (!success) { DYNAMICTOP = ret; ';
-    if (SAFE_HEAP) ret += 'if (asm) { Runtime.setDynamicTop(DYNAMICTOP); }';
-    ret += ' return 0; } }'
+    var ret = '';
+    if (ASSERTIONS) ret += 'assert(DYNAMICTOP_PTR);'; // dynamic area must be ready
```

Comment:
The usage of this function turned out to be a bit more subtle than what I thought. `Runtime.dynamicAlloc` is/can be called in two occassions that I understand:

   - when the static memory area has already been sealed (`staticSealed = true`), but the C runtime has not yet been initialized (`runtimeInitialized=false`), i.e. `_malloc` does not yet own allocating memory within the dynamic memory area.
   - when the static memory area has already been sealed (`staticSealed = true`), and the C runtime has been initialized (`runtimeInitialized=true`), but C `_malloc` was not compiled to the program at all, in which case this function owns dynamic memory allocation throughout the program execution. This is very rare, and looks like only `test_cases` hit this case.

I would have liked to add an assert here that would say `assert(!runtimeInitialized or !(C _malloc has control));` but not quite sure if that is possible to check even. In any case, things would critically break if this function gets called when the compiled `_malloc` is already also serving out memory allocations to the dynamic memory area, but not quite sure how to assert that it doesn't happen.